zkvm/src/signer/mod.rs

```diff
@@ -0,0 +1,68 @@
+#![allow(non_snake_case)]
+
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+
+// Modules for signing protocol
+
+pub mod prover;
+
+pub struct PrivKey(Scalar);
+pub struct PubKey(RistrettoPoint);
+pub struct MultikeyWitness(Vec<PubKey>); // TODO: also include Option<Scalar> for signing key?
+pub struct PubKeyHash(Scalar);
+pub struct Signature {
+    s: Scalar,
+    R: RistrettoPoint,
+}
+
+// TODO: come up with a better name for this!
+pub struct Shared {
+    generator: RistrettoPoint,
+    transcript: Transcript,
+    // X = sum_i ( H(L, X_i) * X_i )
+    agg_pubkey: PubKey,
+    // L = H(X_1 || X_2 || ... || X_n)
+    agg_pubkey_hash: PubKeyHash,
+    // TODO: what should the message representation format be?
+    message: Vec<u8>,
+}
+
+impl MultikeyWitness {
+    pub fn aggregate(&self, transcript: &mut Transcript) -> (PubKey, PubKeyHash) {
+        // L = H(X_1 || X_2 || ... || X_n)
```

Comment:
The nested hash construction is not necessary if we use STROBE. Think of "L" being the internal state of the sponge. What we get with transcript API is a way cleaner "commit all keys, squeeze all randomization factors", without extra intermediate stuff:

<img width="788" alt="image" src="https://user-images.githubusercontent.com/698/53856730-1dd34d00-3f88-11e9-9048-d3f5d1791a89.png">

Cf. https://github.com/interstellar/slingshot/blob/main/zkvm/docs/zkvm-spec.md#aggregated-signature

Reply:
This approach will work for the multikey aggregation; however, when each individual party wants to compute its own `a_i`, then they will have to either have to have all of the individual pubkeys (in order to pass in the pubkeys that are "before" them in ordering to the transcript), or they will have to receive their `a_i` from the coordinator (which seems awkward for coordination). Do you think it's worth the loss in the ergonomic-ness? If so, WDYT is the better approach?

Reply:
For now, I am just implementing this as detailed in the paper (with the exception of not explicitly generating and then committing to "L"... that is done implicitly by feeding the "X_i"s to the transcript"). So, each party feeds its own "X_i" into a clone of the transcript, and gets its own factor_for_key. I think that's the cleanest design, since then parties can get their own factor_for_key without having to know all the other pubkeys.

Reply:
Yeah, seems like cloning transcript for each factor is the way to go to avoid creating all the factors and picking only yours.